src/main/java/net/minecraftforge/client/model/BlockStateLoader.java

```diff
@@ -0,0 +1,686 @@
+            }
+            
+            private Multimap<String, Variant> getPermutations(Map<String, Map<String, Variant>> base)
+            {
+                List<String> sorted = Lists.newArrayList(base.keySet());
+                Collections.sort(sorted);   // Sort to get consistent results.
+                return getPermutations(sorted, base, 0, "", HashMultimap.<String, Variant>create(), null);
+            }
+            
+            private List<Variant> getSubmodelPermutations(Variant baseVar, List<String> sorted, Map<String, List<Variant>> map, int depth, Map<String, Variant> parts, List<Variant> ret)
+            {
+                if (depth >= sorted.size())
+                {
+                    Variant add = new Variant(baseVar); // Create a duplicate variant object so modifying it doesn't modify baseVar.
+                    for (Entry<String, Variant> part : parts.entrySet())    // Put all the parts with single variants for this permutation.
+                        add.submodels.put(part.getKey(), Collections.singletonList(part.getValue()));
+                    ret.add(add);
+                    return ret;
+                }
+                
+                String name = sorted.get(depth);
+                List<Variant> vars = map.get(sorted.get(depth));
+                
+                if (vars != null)
+                {
+                    for (Variant v : vars)
+                    {
+                        if (v != null)
+                        {   // We put this part variant in the permutation's map to add further in recursion, and then remove it afterward just in case.
+                            parts.put(name, v);
+                            getSubmodelPermutations(baseVar, sorted, map, depth + 1, parts, ret);
+                            parts.remove(name);
+                        }
+                    }
+                }
+                else
+                {
+                    getSubmodelPermutations(baseVar, sorted, map, depth + 1, parts, ret);
+                }
+                
+                return ret;
+            }
+            
+            private List<Variant> getSubmodelPermutations(Variant baseVar, Map<String, List<Version1.Variant>> variants)
+            {
+                List<String> sorted = Lists.newArrayList(variants.keySet());
+                Collections.sort(sorted);   // Sort to get consistent results.
+                return getSubmodelPermutations(baseVar, sorted, variants, 0, new HashMap<String, Variant>(), new ArrayList<Variant>());
+            }
+        }
+
+        public static class Variant
+        {
+            public static final Object SET_VALUE = new Object();
+            
+            // ******Set are used to determine which values must be changed to inherit values from a parent.
+            public ResourceLocation model       = null;
+            public boolean          modelSet    = false;
+            public ModelRotation    rotation    = null;
```

Comment:
Would be very nice to be able to support deserealization of custom rotations - ModelRotation is quite limited. Making it wotk with vanilla models is quite a task though.
Also, I'd probably use [Optional](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Optional.html) for most of these fields instead of separate marker bools, since performance isn't critical, but it's also not a critical style choice :P